server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
      * Removes the snapshot deletion from {@link SnapshotDeletionsInProgress} in the cluster state.
      */
-    private void removeSnapshotDeletionFromClusterState(final Collection<SnapshotId> snapshotIds, @Nullable final Exception failure,
-                                                        @Nullable final ActionListener<Void> listener) {
+    private void removeSnapshotDeletionFromClusterState(final SnapshotDeletionsInProgress.Entry deleteEntry,
+                                                        @Nullable final Exception failure, @Nullable final RepositoryData repositoryData) {
+
+        assert failure != null || repositoryData != null : "Either a failure must have occurred or we should have loaded repository data";
+
         clusterService.submitStateUpdateTask("remove snapshot deletion metadata", new ClusterStateUpdateTask() {
+
+            // Snapshots to fail after the state update
+            private final List<Snapshot> snapshotsToFail = new ArrayList<>();
+
+            // Delete uuids to fail because after the state update
+            private final List<String> deletionsToFail = new ArrayList<>();
+
+            // Snapshots that can be finalized after the delete operation has been removed from the cluster state
+            private final List<SnapshotsInProgress.Entry> newFinalizations = new ArrayList<>();
+
             @Override
             public ClusterState execute(ClusterState currentState) {
                 SnapshotDeletionsInProgress deletions = currentState.custom(SnapshotDeletionsInProgress.TYPE);
                 if (deletions != null) {
                     boolean changed = false;
                     if (deletions.hasDeletionsInProgress()) {
-                        assert deletions.getEntries().size() == 1 : "should have exactly one deletion in progress";
-                        SnapshotDeletionsInProgress.Entry entry = deletions.getEntries().get(0);
-                        deletions = deletions.withRemovedEntry(entry);
-                        changed = true;
+                        final SnapshotDeletionsInProgress updatedDeletions = deletions.withRemovedEntry(deleteEntry.uuid());
+                        changed = updatedDeletions != deletions;
+                        deletions = updatedDeletions;
+                        if (failure == null) {
+                            // The delete worked out so we remove the snapshot ids that it removed from the repository from queued up
+                            // delete jobs
+                            deletions = deletions.withRemovedSnapshotIds(deleteEntry.repository(), deleteEntry.getSnapshots());
+                        }
                     }
                     if (changed) {
-                        return ClusterState.builder(currentState).putCustom(SnapshotDeletionsInProgress.TYPE, deletions).build();
+                        // We removed a delete from the cluster state. If that delete caused any snapshots to not have their shard
```

Comment:
This is fairly complex code now unfortunately since we have to deal with assigning new work where possible across queued up snapshots, truncate queued up deletes and handle the complex failure situation of having failed to read repository data before executing the delete.
I did my best to comment reasonably and tried to reuse the shard assignment logic used during snapshot create to keep it simple instead of optimal (see added TODO about optimizing this logic).